exercises/two-bucket/.meta/src/reference/java/TwoBucket.java

```diff
@@ -0,0 +1,139 @@
+import java.util.ArrayList;
+
+class TwoBucket {
+	
+    private class State {
+    	int moves;
+    	int bucketOne;
+    	int bucketTwo;
+    	
+    	public State (int moves, int bucketOne, int bucketTwo) {
+    		this.moves = moves;
+    		this.bucketOne = bucketOne;
+    		this.bucketTwo = bucketTwo;
+    	}
+    	
+    	@Override
+    	public boolean equals(Object o) {
```

Comment:
I'm not too sure if we need to override `equals` here or not, as we don't test it.

If we do need to override `equals` then we might want to override `hashcode` too - otherwise we break the general contract for hashcode! [See here](https://stackoverflow.com/questions/2265503/why-do-i-need-to-override-the-equals-and-hashcode-methods-in-java)

Reply:
Agreed, you don't need to override `equals`. You only need to override it if you want to compare two objects of this class, for example in `assertEquals`. However, as far as I can tell you're not comparing two `State` objects at any point so I don't think it's needed 🙂

Reply:
I only implemented equals because I use the contains method. Won't the contains method call equals?

Reply:
Oh I see! I didn't spot that sorry. Thanks for explaining 🙂 In which case it's fine to override `equals` but I agree with @sjwarner-bp that you should override `hashCode` as well 🙂